blockchain/utxocache.go

```diff
@@ -0,0 +1,816 @@
+	}
+
+	if s.totalMemoryUsage() > threshold {
+		return s.flush(bestState)
+	}
+	return nil
+}
+
+// rollBackBlock rolls back the effects of the block when the state was left in
+// an inconsistent state.  This means that no errors will be raised when the
+// state is invalid.
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) rollBackBlock(block *btcutil.Block, stxos []spentTxOut) error {
+	return disconnectTransactions(s, block, stxos, s)
+}
+
+// rollForwardBlock rolls forward the effects of the block when the state was
+// left in an inconsistent state.  This means that no errors will be raised when
+// the state is invalid.
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) rollForwardBlock(block *btcutil.Block) error {
+	// We don't need the collect stxos and we allow overwriting existing entries.
+	return connectTransactions(s, block, nil, true)
+}
+
+// InitConsistentState checks the consistency status of the utxo state and
+// replays blocks if it lags behind the best state of the blockchain.
+//
+// It needs to be ensured that the chainView passed to this method does not
+// get changed during the execution of this method.
+func (s *utxoCache) InitConsistentState(tip *blockNode, interrupt <-chan struct{}) error {
+	// Load the consistency status from the database.
+	var statusCode byte
+	var statusHash *chainhash.Hash
+	err := s.db.View(func(dbTx database.Tx) error {
+		var err error
+		statusCode, statusHash, err = dbFetchUtxoStateConsistency(dbTx)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	log.Tracef("UTXO cache consistency status from disk: [%d] hash %v",
+		statusCode, statusHash)
+
+	// We can set this variable now already because it will always be valid
+	// unless an error is returned, in which case the state is entirely invalid.
+	// Doing it here prevents forgetting it later.
+	s.lastFlushHash = tip.hash
+
+	// If no status was found, the database is old and didn't have a cached utxo
+	// state yet.  In that case, we set the status to the best state and write
+	// this to the database.
+	if statusCode == ucsEmpty {
+		log.Debugf("Database didn't specify UTXO state consistency: consistent "+
+			"to best chain tip (%v)", tip.hash)
+		err := s.db.Update(func(dbTx database.Tx) error {
+			return dbPutUtxoStateConsistency(dbTx, ucsConsistent, &tip.hash)
+		})
+		return err
+	}
+
+	// If state is consistent, we are done.
+	if statusCode == ucsConsistent && *statusHash == tip.hash {
+		log.Debugf("UTXO state consistent (%d:%v)", tip.height, tip.hash)
+		return nil
+	}
+
+	log.Info("Reconstructing UTXO state after unclean shutdown. This may take " +
+		"a long time...")
+
+	// Even though this should always be true, make sure the fetched hash is in
+	// the best chain.
+	var statusNode *blockNode
+	var statusNodeNext *blockNode // the first one higher than the statusNode
+	for node := tip; node.height > 0; node = node.parent {
```

Comment:
node.height > 0 wont enter this loop if the tip is the genesis block.

Reply:
Suggesting it should be `>= 0`? That doesn't make sense, though, will need another check.

Reply:
This has been fixed. I was debugging yesterday and you are 100% correct.